include/deal.II/particles/particle_handler.h

```diff
@@ -0,0 +1,495 @@
+     * expensive compared to single operations.
+     */
+    void update_cached_numbers();
+
+    /**
+     * Return an iterator to the first particle.
+     */
+    particle_iterator begin() const;
+
+    /**
+     * Return an iterator to the first particle.
+     */
+    particle_iterator begin();
+
+    /**
+     * Return an iterator past the end of the particles.
+     */
+    particle_iterator end() const;
+
+    /**
+     * Return an iterator past the end of the particles.
+     */
+    particle_iterator end();
+
+    /**
+     * Return an iterator to the first particle.
+     */
+    particle_iterator begin_ghost() const;
+
+    /**
+     * Return an iterator to the first particle.
+     */
+    particle_iterator begin_ghost();
+
+    /**
+     * Return an iterator past the end of the particles.
+     */
+    particle_iterator end_ghost() const;
+
+    /**
+     * Return an iterator to the first particle.
+     */
+    particle_iterator end_ghost();
+
+    /**
+     * Return a pair of particle iterators that mark the begin and end of
+     * the particles in a particular cell. The last iterator is the first
+     * particle that is no longer in the cell.
+     */
+    particle_iterator_range
+    particles_in_cell(const typename Triangulation<dim,spacedim>::active_cell_iterator &cell);
+
+
+    /**
+     * Return a pair of particle iterators that mark the begin and end of
+     * the particles in a particular cell. The last iterator is the first
+     * particle that is no longer in the cell.
+     */
+    particle_iterator_range
+    particles_in_cell(const typename Triangulation<dim,spacedim>::active_cell_iterator &cell) const;
+
+    /**
+     * Remove a particle pointed to by the iterator.
+     */
+    void
+    remove_particle(const particle_iterator &particle);
+
+    /**
+     * Insert a particle into the collection of particles. Return an iterator
+     * to the new position of the particle. This function involves a copy of
+     * the particle and its properties. Note that this function is of $O(N \log N)$
+     * complexity for $N$ particles.
+     */
+    particle_iterator
+    insert_particle(const Particle<dim,spacedim> &particle,
+                    const typename Triangulation<dim,spacedim>::active_cell_iterator &cell);
+
+    /**
+     * Insert a number of particles into the collection of particles.
+     * This function involves a copy of the particles and their properties.
+     * Note that this function is of O(n_existing_particles + n_particles) complexity.
+     */
+    void
+    insert_particles(const std::multimap<typename Triangulation<dim,spacedim>::active_cell_iterator,
+                     Particle<dim,spacedim> > &particles);
+
+    /**
+     * Insert a number of particles into the collection of particles.
+     * This function involves a copy of the particles and their properties.
+     * Note that this function is of O(n_existing_particles + n_particles) complexity.
```

Comment:
Maybe add a note that this function expects all locations to be on locally owned cells (or do something in the .cc file when this is not the case, and document what is done in that case).